src/Npgsql/CrateDbDatabaseInfo.cs

```diff
@@ -0,0 +1,235 @@
+                    conn.PostgresParameters.ContainsKey("crate_version")
+                          ? (NpgsqlDatabaseInfo)new CrateDbDatabaseInfo(conn)
+                          : null
+                 );
+    }
+
+    /// <summary>
+    /// Represents a NpgsqlDatabaseInfo-class for usage with CrateDB.
+    /// </summary>
+    public class CrateDbDatabaseInfo : NpgsqlDatabaseInfo
+    {
+        static readonly IDictionary<string, uint> CrateDbBaseTypes = new Dictionary<string, uint>
+        {
+            { "bool", 16 },
+            { "int2", 21 },
+            { "int4", 23 },
+            { "json", 114 },
+            { "float4", 700 },
+            { "char", 18 },
+            { "int8", 20 },
+            { "timestamptz", 1184 },
+            { "float8", 701 },
+            { "varchar", 1043 }
+        };
+
+        static readonly IDictionary<string, uint> CrateDbArrayTypes = new Dictionary<string, uint>
+        {
+            { "_bool", 1000 },
+            { "_int2", 1005 },
+            { "_int4", 1007 },
+            { "_json", 199 },
+            { "_float4", 1021 },
+            { "_char", 1002 },
+            { "_int8", 1016 },
+            { "_timestamptz", 1185 },
+            { "_float8", 1022 },
+            { "_varchar", 1015 }
+        };
+
+        static readonly IDictionary<string, string> PgTypeNameToInternalName = new Dictionary<string, string>
+        {
+            { "boolean", "bool" },
+            { "smallint", "int2" },
+            { "integer", "int4" },
+            { "real", "float4" },
+            { "bigint", "int8" },
+            { "timestamp with time zone", "timestamptz" },
+            { "double precision", "float8" },
+            { "character varying", "varchar" }
+        };
+
+        readonly NpgsqlConnection conn;
+
+        /// <summary>
+        /// Creates an instance of the CrateDbDatabaseInfo class.
+        /// </summary>
+        /// <param name="conn"></param>
+        public CrateDbDatabaseInfo(NpgsqlConnection conn)
+        {
+            this.conn = conn;
+
+            if (Version.TryParse(conn.PostgresParameters["crate_version"], out var v))
+            {
+                Version = v;
+            }
+
+            HasIntegerDateTimes = conn.PostgresParameters.TryGetValue("integer_datetimes", out var intDateTimes) &&
+                intDateTimes == "on";
+        }
+
+        /// <summary>
+        /// Returns the data types supported by CrateDB.
+        /// </summary>
+        /// <returns></returns>
+        protected override IEnumerable<PostgresType> GetTypes()
+        {
+            IEnumerable<PostgresType> baseTypes = CrateDbBaseTypes.Select(p => new CrateDbBaseType(p.Key, p.Value)).ToList();
+            IEnumerable<PostgresType> arrayTypes = CrateDbArrayTypes.Select(p => new CrateDbArrayType(p.Key, p.Value, baseTypes.FirstOrDefault(t => string.Equals(t.InternalName, p.Key.Substring(1)))));
+
+            return baseTypes.Concat(arrayTypes);
+        }
```

Comment:
You could just `yield` these back here and save the allocations. Though, it might not make a difference in the grand scheme of things. Feel free to disregard if there's a compelling reason not to `yield` in this case.

Reply:
i refactored the implementation. please check it. The first allocation and toList() is needed cause in the second Enumerable we access the data from the first list.